File: Integraupt-Backend/BackendAdReserva/src/main/java/com/AdReserva/service/AdminReservaService.java
package com.AdReserva.service;

import com.AdReserva.dto.AdminGestionReservaRequest;
import com.AdReserva.dto.AdminReservaCardDto;
import com.AdReserva.dto.AdminReservaFilter;
import com.AdReserva.dto.AdminReservaFiltersResponse;
import com.AdReserva.dto.AdminReservaListResponse;
import com.AdReserva.dto.AdminReservaSummaryDto;
import com.AdReserva.dto.EscuelaOptionDto;
import com.AdReserva.dto.ReservaGestionAccion;
import com.AdReserva.dto.SimpleOptionDto;
import com.AdReserva.model.Administrativo;
import com.AdReserva.model.Facultad;
import com.AdReserva.model.Escuela;
import com.AdReserva.model.Reserva;
import com.AdReserva.model.ReservaGestion;
import com.AdReserva.repository.AdministrativoRepository;
import com.AdReserva.repository.EscuelaRepository;
import com.AdReserva.repository.EspacioRepository;
import com.AdReserva.repository.FacultadRepository;
import com.AdReserva.repository.ReservaGestionRepository;
import com.AdReserva.repository.ReservaRepository;
import com.AdReserva.repository.projection.AdminReservaEstadoProjection;
import com.AdReserva.repository.projection.AdminReservaProjection;
import com.AdReserva.service.mapper.AdminReservaMapper;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDate;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class AdminReservaService {

    private final ReservaRepository reservaRepository;
    private final ReservaGestionRepository reservaGestionRepository;
    private final FacultadRepository facultadRepository;
    private final EscuelaRepository escuelaRepository;
    private final EspacioRepository espacioRepository;
    private final AdministrativoRepository administrativoRepository;

    public AdminReservaService(
            ReservaRepository reservaRepository,
            ReservaGestionRepository reservaGestionRepository,
            FacultadRepository facultadRepository,
            EscuelaRepository escuelaRepository,
            EspacioRepository espacioRepository,
            AdministrativoRepository administrativoRepository
    ) {
        this.reservaRepository = reservaRepository;
        this.reservaGestionRepository = reservaGestionRepository;
        this.facultadRepository = facultadRepository;
        this.escuelaRepository = escuelaRepository;
        this.espacioRepository = espacioRepository;
        this.administrativoRepository = administrativoRepository;
    }

    @Transactional(readOnly = true)
    public AdminReservaListResponse obtenerReservas(AdminReservaFilter filter) {
        String estado = filter.estadoOpt().map(this::normalizarEstado).orElse(null);
        String tipoEspacio = filter.tipoEspacioOpt().map(this::normalizarTipo).orElse(null);
        Integer facultadId = filter.facultadIdOpt().orElse(null);
        Integer escuelaId = filter.escuelaIdOpt().orElse(null);
        LocalDate fecha = filter.fechaReservaOpt().orElse(null);
        String search = filter.searchOpt().map(String::trim).map(String::toLowerCase).orElse(null);

        if (esSupervisor(filter.rolOpt())) {
            Integer usuarioId = filter.usuarioIdOpt().orElseThrow(() ->
                    new ResponseStatusException(HttpStatus.BAD_REQUEST,
                            "Se requiere el identificador del usuario supervisor."));
            SupervisorAssignment assignment = obtenerAsignacionSupervisor(usuarioId);
            facultadId = assignment.facultadId();
            escuelaId = assignment.escuelaId();
        }


        List<AdminReservaCardDto> reservas = reservaRepository.buscarReservasParaAdmin(
                        estado, tipoEspacio, facultadId, escuelaId, fecha, search)
                .stream()
                .map(AdminReservaMapper::toDto)
                .toList();

        AdminReservaSummaryDto resumen = construirResumen(
                reservaRepository.obtenerResumenEstados(tipoEspacio, facultadId, escuelaId, fecha, search)
        );

        return new AdminReservaListResponse(reservas, resumen);
    }

    @Transactional(readOnly = true)
    public AdminReservaFiltersResponse obtenerFiltros(Integer usuarioId, String rol) {
        List<String> tipos = espacioRepository.obtenerTiposDeEspacio();

        if (esSupervisor(rol)) {
            if (usuarioId == null) {
                throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                        "Se requiere el identificador del usuario supervisor.");
            }
            SupervisorAssignment assignment = obtenerAsignacionSupervisor(usuarioId);

            List<SimpleOptionDto> facultades = List.of(
                    new SimpleOptionDto(assignment.facultadId(), assignment.facultadNombre())
            );

            List<EscuelaOptionDto> escuelas = List.of(
                    new EscuelaOptionDto(assignment.escuelaId(), assignment.escuelaNombre(), assignment.facultadId())
            );

            return new AdminReservaFiltersResponse(tipos, facultades, escuelas);
        }

        List<SimpleOptionDto> facultades = facultadRepository.findAll().stream()
                .sorted(Comparator.comparing(Facultad::getNombre, String.CASE_INSENSITIVE_ORDER))
                .map(facultad -> new SimpleOptionDto(facultad.getId(), facultad.getNombre()))
                .toList();

        List<EscuelaOptionDto> escuelas = escuelaRepository.findAll().stream()
                .sorted(Comparator.comparing(Escuela::getNombre, String.CASE_INSENSITIVE_ORDER))
                .map(escuela -> new EscuelaOptionDto(escuela.getId(), escuela.getNombre(), escuela.getFacultadId()))
                .toList();

        return new AdminReservaFiltersResponse(tipos, facultades, escuelas);
    }

    @Transactional
    public AdminReservaCardDto gestionarReserva(Integer reservaId, AdminGestionReservaRequest request) {
        Reserva reserva = reservaRepository.findById(reservaId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "La reserva solicitada no existe"));

        ReservaGestionAccion accion;
        try {
            accion = ReservaGestionAccion.fromValue(request.accion());
        } catch (IllegalArgumentException ex) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, ex.getMessage());
        }

        if (Objects.equals(reserva.getEstado(), accion.estadoResultado())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "La reserva ya se encuentra en el estado solicitado");
        }

        reserva.setEstado(accion.estadoResultado());
        reservaRepository.save(reserva);

        ReservaGestion gestion = new ReservaGestion();
        gestion.setReservaId(reserva.getId());
        gestion.setUsuarioGestionId(request.usuarioGestionId());
        gestion.setAccion(accion.accionBd());
        gestion.setMotivo(request.motivo().trim());
        gestion.setComentarios(request.comentarios() != null ? request.comentarios().trim() : null);
        reservaGestionRepository.save(gestion);

        AdminReservaProjection projection = reservaRepository.buscarReservaPorId(reserva.getId());
        if (projection == null) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
                    "No fue posible obtener la información actualizada de la reserva");
        }
        return AdminReservaMapper.toDto(projection);
    }

    private AdminReservaSummaryDto construirResumen(List<AdminReservaEstadoProjection> datos) {
        if (datos == null || datos.isEmpty()) {
            return AdminReservaSummaryDto.empty();
        }
        Map<String, Long> porEstado = datos.stream()
                .collect(Collectors.toMap(AdminReservaEstadoProjection::getEstado, AdminReservaEstadoProjection::getTotal));

        return new AdminReservaSummaryDto(
                porEstado.getOrDefault("Pendiente", 0L),
                porEstado.getOrDefault("Aprobada", 0L),
                porEstado.getOrDefault("Rechazada", 0L),
                porEstado.getOrDefault("Cancelada", 0L)
        );
    }

    private boolean esSupervisor(String rol) {
        return rol != null && rol.trim().equalsIgnoreCase("supervisor");
    }

    private boolean esSupervisor(Optional<String> rolOpt) {
        return rolOpt.map(this::esSupervisor).orElse(false);
    }

    private SupervisorAssignment obtenerAsignacionSupervisor(Integer usuarioId) {
        Administrativo administrativo = administrativoRepository.findByUsuarioId(usuarioId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.FORBIDDEN,
                        "No se encontró un perfil administrativo para el usuario especificado."));

        Integer escuelaAsignadaId = administrativo.getEscuelaId();
        if (escuelaAsignadaId == null) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN,
                    "El supervisor no tiene una escuela asignada.");
        }

        Escuela escuela = escuelaRepository.findById(escuelaAsignadaId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "La escuela asignada al supervisor no existe."));

        Facultad facultad = facultadRepository.findById(escuela.getFacultadId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "La facultad asociada a la escuela del supervisor no existe."));

        return new SupervisorAssignment(
                escuela.getId(),
                escuela.getNombre(),
                facultad.getId(),
                facultad.getNombre()
        );
    }

    private String normalizarEstado(String estado) {
        if (estado == null) {
            return null;
        }
        return switch (estado.trim().toLowerCase(Locale.ROOT)) {
            case "pendiente" -> "Pendiente";
            case "aprobada" -> "Aprobada";
            case "rechazada" -> "Rechazada";
            case "cancelada" -> "Cancelada";
            default -> throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "Estado de reserva desconocido: " + estado);
        };
    }

    private String normalizarTipo(String tipo) {
        if (tipo == null) {
            return null;
        }
        String sanitized = tipo.trim();
        if (sanitized.isEmpty()) {
            return null;
        }
        return sanitized.substring(0, 1).toUpperCase(Locale.ROOT) + sanitized.substring(1).toLowerCase(Locale.ROOT);
    }

    private record SupervisorAssignment(Integer escuelaId, String escuelaNombre,
                                        Integer facultadId, String facultadNombre) {
    }
}

File: Integraupt-Backend/BackendAdReserva/src/Services/AdminReservaService.php
<?php

namespace App\Services;

use App\Core\HttpException;
use App\DTO\AdminGestionReservaRequest;
use App\DTO\AdminReservaCardDto;
use App\DTO\AdminReservaFilter;
use App\DTO\AdminReservaFiltersResponse;
use App\DTO\AdminReservaListResponse;
use App\DTO\AdminReservaSummaryDto;
use App\DTO\EscuelaOptionDto;
use App\DTO\SimpleOptionDto;
use App\Models\ReservaGestion;
use App\Repositories\AdministrativoRepository;
use App\Repositories\EscuelaRepository;
use App\Repositories\EspacioRepository;
use App\Repositories\FacultadRepository;
use App\Repositories\ReservaGestionRepository;
use App\Repositories\ReservaRepository;

class AdminReservaService {
    private ReservaRepository $reservaRepository;
    private ReservaGestionRepository $reservaGestionRepository;
    private FacultadRepository $facultadRepository;
    private EscuelaRepository $escuelaRepository;
    private EspacioRepository $espacioRepository;
    private AdministrativoRepository $administrativoRepository;

    public function __construct() {
        $this->reservaRepository = new ReservaRepository();
        $this->reservaGestionRepository = new ReservaGestionRepository();
        $this->facultadRepository = new FacultadRepository();
        $this->escuelaRepository = new EscuelaRepository();
        $this->espacioRepository = new EspacioRepository();
        $this->administrativoRepository = new AdministrativoRepository();
    }

    public function obtenerReservas(AdminReservaFilter $filter): AdminReservaListResponse {
        $estado = $filter->getEstado() !== null ? $this->normalizarEstado($filter->getEstado()) : null;
        $tipoEspacio = $filter->getTipoEspacio() !== null ? $this->normalizarTipo($filter->getTipoEspacio()) : null;
        $facultadId = $filter->getFacultadId();
        $escuelaId = $filter->getEscuelaId();
        $fecha = $filter->getFechaReserva();
        $search = $filter->getSearch() !== null ? trim($filter->getSearch()) : null;

        if ($this->esSupervisor($filter->getRol())) {
            $usuarioId = $filter->getUsuarioId();
            if ($usuarioId === null) {
                throw new HttpException("Se requiere el identificador del usuario supervisor.", 400);
            }
            $assignment = $this->obtenerAsignacionSupervisor($usuarioId);
            $facultadId = $assignment['facultadId'];
            $escuelaId = $assignment['escuelaId'];
        }

        $reservas = $this->reservaRepository->buscarReservasParaAdmin(
            $estado,
            $tipoEspacio,
            $facultadId,
            $escuelaId,
            $fecha,
            $search
        );

        $datosResumen = $this->reservaRepository->obtenerResumenEstados(
            $tipoEspacio,
            $facultadId,
            $escuelaId,
            $fecha,
            $search
        );

        $resumen = $this->construirResumen($datosResumen);

        return new AdminReservaListResponse($reservas, $resumen);
    }

    public function obtenerFiltros(?int $usuarioId, ?string $rol): AdminReservaFiltersResponse {
        $tipos = $this->espacioRepository->obtenerTiposDeEspacio();

        if ($this->esSupervisor($rol)) {
            if ($usuarioId === null) {
                throw new HttpException("Se requiere el identificador del usuario supervisor.", 400);
            }
            $assignment = $this->obtenerAsignacionSupervisor($usuarioId);

            $facultades = [
                new SimpleOptionDto($assignment['facultadId'], $assignment['facultadNombre'])
            ];

            $escuelas = [
                new EscuelaOptionDto($assignment['escuelaId'], $assignment['escuelaNombre'], $assignment['facultadId'])
            ];

            return new AdminReservaFiltersResponse($tipos, $facultades, $escuelas);
        }

        // Obtener y ordenar facultades alfabéticamente (insensible a mayúsculas/minúsculas)
        $facultadesDb = $this->facultadRepository->findAll();
        usort($facultadesDb, function ($a, $b) {
            return strcasecmp($a->nombre, $b->nombre);
        });
        $facultades = array_map(function ($facultad) {
            return new SimpleOptionDto($facultad->id, $facultad->nombre);
        }, $facultadesDb);

        // Obtener y ordenar escuelas alfabéticamente
        $escuelasDb = $this->escuelaRepository->findAll();
        usort($escuelasDb, function ($a, $b) {
            return strcasecmp($a->nombre, $b->nombre);
        });
        $escuelas = array_map(function ($escuela) {
            return new EscuelaOptionDto($escuela->id, $escuela->nombre, $escuela->facultadId);
        }, $escuelasDb);

        return new AdminReservaFiltersResponse($tipos, $facultades, $escuelas);
    }

    public function gestionarReserva(int $reservaId, AdminGestionReservaRequest $request): AdminReservaCardDto {
        $reserva = $this->reservaRepository->findById($reservaId);
        if ($reserva === null) {
            throw new HttpException("La reserva solicitada no existe", 404);
        }

        $accion = $request->accion;
        if ($accion === null || trim($accion) === '') {
            throw new HttpException("La acción de gestión no puede ser nula", 400);
        }

        $accionSanitizada = strtolower(trim($accion));
        if ($accionSanitizada === 'aprobar') {
            $accionBd = "Aprobar";
            $estadoResultado = "Aprobada";
        } elseif ($accionSanitizada === 'rechazar') {
            $accionBd = "Rechazar";
            $estadoResultado = "Rechazada";
        } else {
            throw new HttpException("Acción de gestión desconocida: " . $accion, 400);
        }

        if ($reserva->estado === $estadoResultado) {
            throw new HttpException("La reserva ya se encuentra en el estado solicitado", 409);
        }

        // Actualizar el estado de la reserva
        $reserva->estado = $estadoResultado;
        $this->reservaRepository->save($reserva);

        // Guardar registro de la gestión
        $gestion = new ReservaGestion(
            null,
            $reserva->id,
            $request->usuarioGestionId,
            null, // Usará por defecto date('Y-m-d H:i:s') en su constructor
            $accionBd,
            trim($request->motivo),
            $request->comentarios !== null ? trim($request->comentarios) : null
        );
        $this->reservaGestionRepository->save($gestion);

        // Obtener la reserva actualizada y mapeada
        $updatedCard = $this->reservaRepository->buscarReservaPorId($reserva->id);
        if ($updatedCard === null) {
            throw new HttpException("No fue posible obtener la información actualizada de la reserva", 500);
        }

        return $updatedCard;
    }

    private function construirResumen(array $datos): AdminReservaSummaryDto {
        if (empty($datos)) {
            return AdminReservaSummaryDto::empty();
        }

        return new AdminReservaSummaryDto(
            $datos['Pendiente'] ?? 0,
            $datos['Aprobada'] ?? 0,
            $datos['Rechazada'] ?? 0,
            $datos['Cancelada'] ?? 0
        );
    }

    private function esSupervisor(?string $rol): bool {
        return $rol !== null && strtolower(trim($rol)) === 'supervisor';
    }

    private function obtenerAsignacionSupervisor(int $usuarioId): array {
        $administrativo = $this->administrativoRepository->findByUsuarioId($usuarioId);
        if ($administrativo === null) {
            throw new HttpException("No se encontró un perfil administrativo para el usuario especificado.", 403);
        }

        $escuelaAsignadaId = $administrativo->escuelaId;
        if ($escuelaAsignadaId === null) {
            throw new HttpException("El supervisor no tiene una escuela asignada.", 403);
        }

        $escuela = $this->escuelaRepository->findById($escuelaAsignadaId);
        if ($escuela === null) {
            throw new HttpException("La escuela asignada al supervisor no existe.", 404);
        }

        $facultad = $this->facultadRepository->findById($escuela->facultadId);
        if ($facultad === null) {
            throw new HttpException("La facultad asociada a la escuela del supervisor no existe.", 404);
        }

        return [
            'escuelaId' => $escuela->id,
            'escuelaNombre' => $escuela->nombre,
            'facultadId' => $facultad->id,
            'facultadNombre' => $facultad->nombre
        ];
    }

    private function normalizarEstado(string $estado): string {
        $clean = strtolower(trim($estado));
        return match ($clean) {
            'pendiente' => 'Pendiente',
            'aprobada' => 'Aprobada',
            'rechazada' => 'Rechazada',
            'cancelada' => 'Cancelada',
            default => throw new HttpException("Estado de reserva desconocido: " . $estado, 400)
        };
    }

    private function normalizarTipo(string $tipo): ?string {
        $sanitized = trim($tipo);
        if ($sanitized === '') {
            return null;
        }
        // Capitalización de la primera letra (ej. "laboratorio" -> "Laboratorio")
        return mb_strtoupper(mb_substr($sanitized, 0, 1)) . mb_strtolower(mb_substr($sanitized, 1));
    }
}
